### PR TITLE
metadata refactors

### LINKED ISSUE #1151: metadata
+ `deleted` -> `is_deleted` to indicate that it's a boolean

### LINKED ISSUE #1216: refactor: metadata signatures

### LINKED ISSUE #1217: filename keys
filenames should be encrypted with their key, and hmac'd with their parent's key.

currently the parent key is used for both operations.

### LINKED ISSUE #1219: digests in metadatas
currently we store digests in their own table in core, we should store this in the metadata, so that you can have a cryptographically verifiable statement of who edited this last and whether they were permissioned to do so (read v write access).

### LINKED ISSUE #1220: model tree validations better

### PR DESCRIPTION
**Pull request by Parth**
(wip)

fixes #1151 
fixes #1216 
fixes #1220
fixes #1219
fixes #1217

Post PR todo's:

+ [ ] new tests around prune and merge

### CODE REVIEW

File: clients/android/app/src/main/java/app/lockbook/util/CoreData.kt [hunk 1]
```diff
@@ -6,22 +6,24 @@ import kotlinx.serialization.SerialName
 import kotlinx.serialization.Serializable
 import kotlinx.serialization.json.JsonClassDiscriminator
 
+//    pub id: Uuid,
+//    pub parent: Uuid,
+//    pub name: String,
+//    pub file_type: FileType,
+//    pub last_modified: u64,
+//    pub last_modified_by: Username,
+
```
Comment (steverusso) [inline clients/android/app/src/main/java/app/lockbook/util/CoreData.kt:15]: Can this get dropped now?

File: core/Cargo.toml [hunk 2]
```diff
@@ -41,12 +40,11 @@ libsecp256k1 = "0.7.1"
 tracing = "0.1.5"
 tracing-subscriber = "0.3.9"
 tracing-appender = "0.2"
-hmdb = "0.1.12"
+hmdb = { path = "../../../hmdb" }
```
Comment (steverusso) [inline core/Cargo.toml:43]: Reminder that this should probably be reverted (or discussed?) before merge

File: core/libs/shared/Cargo.toml [hunk 3]
```diff
@@ -12,6 +12,16 @@ serde = { version = "1.0", features = ["derive"] }
 serde_bytes = "0.11"
 libsecp256k1 = "0.7.1"
 uuid = { version = "0.8.1", features = ["v4", "serde"] }
+aead = "0.4.2"
+aes-gcm = "0.9.3"
+bincode = "1.2.1"
+sha2 = "0.9.0"
+hmac = "0.11.0"
+hmdb = { path = "../../../../../hmdb" }
```
Comment (steverusso) [inline core/libs/shared/Cargo.toml:20]: Reminder for same here

File: core/libs/shared/src/file_metadata.rs [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+            folder_access_keys: symkey::encrypt(parent_key, &key)?,
+        })
+    }
+
+    pub fn sign(self, account: &Account) -> SharedResult<SignedFile> {
+        pubkey::sign(&account.private_key, self, get_time)
+    }
+}
+
+impl PartialEq for FileMetadata {
```
Comment (steverusso) [inline core/libs/shared/src/file_metadata.rs:77]: Why can't this be derived? If there's a reason, I think we should state it in a doc comment right here.

File: core/libs/shared/src/file_metadata.rs [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    pub old: Option<SignedFile>,
+    pub new: SignedFile,
+}
+
+impl fmt::Debug for FileDiff {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut result = &mut f.debug_struct("FileDiff");
+        result = result.field("id", self.id());
+        for diff in self.diff() {
+            match diff {
```
Comment (steverusso) [inline core/libs/shared/src/file_metadata.rs:139]:
```
result = match diff {
    ...
};
```
?

Comment (steverusso) [reply core/libs/shared/src/file_metadata.rs]: https://github.com/lockbook/lockbook/pull/1221/commits/7c718f2eb9e5f688202e4b64bdf827a49601845f

File: core/libs/shared/src/file_metadata.rs (hunk 4 above)
Comment (Parth) [reply core/libs/shared/src/file_metadata.rs:76]: https://github.com/lockbook/lockbook/issues/1251

File: clients/android/app/src/main/java/app/lockbook/util/CoreData.kt (hunk 1 above)
Comment (steverusso) [reply clients/android/app/src/main/java/app/lockbook/util/CoreData.kt:15]: @Parth

**Review by tvanderstad — APPROVED** [review]

**Pull request merged**